sms_alternative_provider/models/ir_sms_gateway.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2024 Hunki Enterprises BV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl-3.0)
+
+import operator
+
+from odoo import _, api, exceptions, fields, models
+
+
+class IrSmsGateway(models.Model):
+
+    _name = "ir.sms.gateway"
+    _order = "sequence"
+    _desc = "SMS gateway provider"
+
+    name = fields.Char(required=True)
+    active = fields.Boolean(default=True)
+    gateway_type = fields.Selection([("iap", "Odoo IAP")], string="Type", required=True)
+    sequence = fields.Integer(default=lambda self: self._default_sequence())
+    prefix = fields.Char(
+        help="Fill in for which phone number prefix(es) this provider is used, ie "
+        "'+31' to use it for NL, '+31 +49' to use it for DE and NL, or '+3120' to use "
+        "it for Amsterdam/NL. Leave empty to send to any number.",
+    )
+    description = fields.Html(compute="_compute_description")
+    company_id = fields.Many2one("res.company")
+
+    def _default_sequence(self):
+        return (
+            max(list(filter(None, self.sudo().search([]).mapped("sequence"))) or [0])
+            + 1
+        )
+
+    @api.depends("gateway_type")
+    def _compute_description(self):
+        """
+        Call a provider specific function _get_description_$gateway_type if defined
+        to show a description/help text on the gateway form
+        """
+        for this in self:
+            this.description = getattr(
+                this, "_get_description_%s" % this.gateway_type, lambda: False
+            )()
+
+    # SMS sending functions
+
+    def send(self, messages):
+        """
+        Send a list of SMS via the current provider
+
+        Return list of dictionaries [{
+            'id': sms.sms id,
+            'state': sms.sms#state,
+            'failure_type': sms.sms#failure_type
+        }]
+        """
+        self.ensure_one()
+        result = getattr(self, "_send_%s" % self.gateway_type)(messages) or []
+        return [dict(result_dict, sms_gateway_id=self.id) for result_dict in result]
+
+    @api.model
+    def _send(self, messages, handle_results=True, raise_exception=True):
+        """
+        Select provider(s) based on messages, call their provider specific function
+        _send_$gateway_type to actually send SMS
+
+        messages is a list of dictionaries [{
+            'id': sms.sms id,
+            'number': phone number,
+            'content': sms content,
```

Comment:
Discussion point - some providers, not Odoo IAP I think but others, provide the possibility to specify a FROM number. Originally this can come either from a default specified on the account, or from the sms template. But in the end it has to end up in this dictionary. What would be the best way of supporting that?

Reply:
the default on the account is easy, as the provider-specific send method is called on the record of the account. The messagebird implementation I added here takes this from the company's name.

I don't see why we need this to be passed in this dict at all?

It will be difficult to do this on the template level, because in general an sms doesn't know which template it comes from (see the hack I'm using to solve this for sms_masked_content)

Reply:
Well, let's leave it for now. If I really need it I could always do a custom override to pass the sms template or the from-number as an extra provider-specific parameter and hack it in somehow. And then propose a generic fix, if needed. For now I'm setting it in my custom module via the sms account.